controllers/remote/keyedmutex.go

```diff
@@ -0,0 +1,78 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package remote
+
+import "sync"
+
+// keyedMutex is a mutex locking on the key provided to the Lock function.
+// Only one caller can hold the lock for a specific key at a time.
+// A second Lock call if the lock is already held for a key returns false.
+type keyedMutex struct {
+	locksMtx sync.Mutex
+	locks    map[interface{}]*sync.Mutex
+}
+
+// newKeyedMutex creates a new keyed mutex ready for use.
+func newKeyedMutex() *keyedMutex {
+	return &keyedMutex{
+		locks: make(map[interface{}]*sync.Mutex),
+	}
+}
+
+// unlock unlocks a currently locked key.
+type unlock func()
+
+// TryLock locks the passed in key if it's not already locked.
+// Returns the unlock function to release the lock on the key.
+// A second Lock call if the lock is already held for a key returns false.
+// In the ClusterCacheTracker case the key is the ObjectKey for a cluster.
+func (k *keyedMutex) TryLock(key interface{}) (unlock, bool) {
```

Comment:
Instead of returning a function, could we return a (bool, error) and pass in a function as the second parameter that runs if an only if the lock can be held?

Reply:
I refactored the whole thing in another way.

Basically it now works like a normal Mutex with .Lock and .Unlock functions (with the only difference that an objectKey has to be passed in in both).